File: indobase-helpdesk/bridge/src/helpdesk-map.ts
/**
 * Deterministic org/project → Helpdesk team / queue keys.
 *
 * Must stay in sync with `frappe-app/indobase_helpdesk/indobase_helpdesk/utils/helpdesk_map.py`
 * and `apps/studio/lib/api/saas/helpdesk-launch-shared.ts`.
 */

export type HelpdeskScopeMap = {
  orgSlug: string
  projectRef: string
  teamKey: string
  queueKey: string
  teamTitle: string
  queueTitle: string
}

const MAX_KEY_LEN = 64

function cleanSlug(input: string): string {
  return input
    .toLowerCase()
    .split('')
    .filter((c) => /[a-z0-9-]/.test(c))
    .join('')
    .slice(0, MAX_KEY_LEN)
}

function cleanProjectRef(input: string): string {
  return input
    .toLowerCase()
    .split('')
    .filter((c) => /[a-z0-9]/.test(c))
    .join('')
    .slice(0, 40)
}

/** Stable support team key for an Indobase organization slug. */
export function helpdeskTeamKeyForOrgSlug(orgSlug: string): string {
  const cleaned = cleanSlug(orgSlug)
  if (!cleaned) return 'ib-hd-org-default'
  return `ib-hd-org-${cleaned}`.slice(0, MAX_KEY_LEN)
}

/** Stable ticket queue key for an Indobase project ref. */
export function helpdeskQueueKeyForProjectRef(projectRef: string): string {
  const cleaned = cleanProjectRef(projectRef)
  if (!cleaned) return 'ib-hd-proj-default'
  return `ib-hd-proj-${cleaned}`.slice(0, MAX_KEY_LEN)
}

export function buildHelpdeskScopeMap(opts: {
  orgSlug: string
  projectRef: string
  projectName?: string
  organizationName?: string
}): HelpdeskScopeMap {
  const orgSlug = opts.orgSlug.trim()
  const projectRef = opts.projectRef.trim()
  const teamKey = helpdeskTeamKeyForOrgSlug(orgSlug)
  const queueKey = helpdeskQueueKeyForProjectRef(projectRef)
  const teamTitle = (opts.organizationName || orgSlug || 'Organization').slice(0, 140)
  const queueTitle = (opts.projectName || projectRef || 'Project').slice(0, 140)

  return {
    orgSlug,
    projectRef,
    teamKey,
    queueKey,
    teamTitle,
    queueTitle,
  }
}

/** Bridge deep link after SSO — agent desk for the project queue. */
export function helpdeskAgentPath(map: HelpdeskScopeMap): string {
  return `/h/${encodeURIComponent(map.teamKey)}/${encodeURIComponent(map.queueKey)}`
}

/** Customer portal entry for the same org/project scope. */
export function helpdeskPortalPath(map: HelpdeskScopeMap): string {
  return `/portal/${encodeURIComponent(map.teamKey)}/${encodeURIComponent(map.queueKey)}`
}

/** Upstream Frappe Helpdesk SPA entry (production proxy target). */
export function upstreamHelpdeskPath(bridgePath: string, portal = false): string {
  if (bridgePath.startsWith('/h/')) {
    const parts = bridgePath.split('/').filter(Boolean)
    if (parts.length >= 3) {
      const queueKey = decodeURIComponent(parts[2])
      return `/helpdesk?ib_queue=${encodeURIComponent(queueKey)}`
    }
  }
  if (bridgePath.startsWith('/portal/')) {
    const parts = bridgePath.split('/').filter(Boolean)
    if (parts.length >= 3) {
      const queueKey = decodeURIComponent(parts[2])
      return `/helpdesk/my-tickets?ib_queue=${encodeURIComponent(queueKey)}`
    }
    return '/helpdesk/my-tickets'
  }
  if (bridgePath.startsWith('/helpdesk')) return bridgePath
  return portal ? '/helpdesk/my-tickets' : '/helpdesk'
}

File: indobase-helpdesk/frappe-app/indobase_helpdesk/indobase_helpdesk/utils/helpdesk_map.py
"""Deterministic org/project → Helpdesk team / queue keys (mirror bridge/src/helpdesk-map.ts)."""

from __future__ import annotations

MAX_KEY_LEN = 64


def _clean_slug(value: str) -> str:
	cleaned = "".join(c for c in value.lower() if c.isalnum() or c == "-")
	return cleaned[:MAX_KEY_LEN]


def _clean_project_ref(value: str) -> str:
	cleaned = "".join(c for c in value.lower() if c.isalnum())
	return cleaned[:40]


def helpdesk_team_key_for_org_slug(org_slug: str) -> str:
	cleaned = _clean_slug(org_slug or "")
	if not cleaned:
		return "ib-hd-org-default"
	return f"ib-hd-org-{cleaned}"[:MAX_KEY_LEN]


def helpdesk_queue_key_for_project_ref(project_ref: str) -> str:
	cleaned = _clean_project_ref(project_ref or "")
	if not cleaned:
		return "ib-hd-proj-default"
	return f"ib-hd-proj-{cleaned}"[:MAX_KEY_LEN]


def build_helpdesk_scope_map(
	*,
	org_slug: str,
	project_ref: str,
	project_name: str | None = None,
	organization_name: str | None = None,
) -> dict[str, str]:
	team_key = helpdesk_team_key_for_org_slug(org_slug)
	queue_key = helpdesk_queue_key_for_project_ref(project_ref)
	return {
		"org_slug": (org_slug or "").strip(),
		"project_ref": (project_ref or "").strip(),
		"team_key": team_key,
		"queue_key": queue_key,
		"team_title": (organization_name or org_slug or "Organization")[:140],
		"queue_title": (project_name or project_ref or "Project")[:140],
	}


def helpdesk_agent_path(scope_map: dict[str, str]) -> str:
	return f"/h/{scope_map['team_key']}/{scope_map['queue_key']}"


def helpdesk_portal_path(scope_map: dict[str, str]) -> str:
	return f"/portal/{scope_map['team_key']}/{scope_map['queue_key']}"
